### PR TITLE
Jep/table view

### PR DESCRIPTION
**Pull request by Drakae**
Added a table webview that could be used for the simulation variables

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -8,11 +8,13 @@
             "name": "Launch VS Code Extension",
             "type": "extensionHost",
             "request": "launch",
-            "args": ["--extensionDevelopmentPath=${workspaceFolder}/keith-vscode"],
+            "args": [
+                "--extensionDevelopmentPath=${workspaceFolder}/keith-vscode",
+				"--extensionDevelopmentPath=${workspaceFolder}/../klighd-vscode/applications/klighd-vscode"
```
Comment (Eddykasp) [inline .vscode/launch.json:13]: this is use case specific and shouldn't generally be like this

File: keith-vscode/package.json [hunk 2]
```diff
@@ -561,8 +561,9 @@
     "lint": "eslint .",
     "build": "webpack --mode production",
     "watch": "webpack --watch",
-    "package": "vsce package --yarn -o keith-vscode.vsix",
-    "vscode:prepublish": "yarn run build"
+    "package": "vsce package --yarn -o keith-vscode.vsix && yarn run copy-webview",
+    "vscode:prepublish": "yarn run build",
+    "copy-webview": "xcopy /y ..\\table-webview\\out\\* dist"
```
Comment (Eddykasp) [inline keith-vscode/package.json:566]: what exactly does this do?

File: keith-vscode/src/constants.ts [hunk 3]
```diff
@@ -13,6 +13,10 @@ export const views = {
         id: 'kieler-simulation-tree',
         name: 'KIELER Simulation Tree',
     },
+    checker: {
+        id: 'kieler-model-checker',
+        name: 'KIELER Model Checker',
+    }
```
Comment (Eddykasp) [inline keith-vscode/src/constants.ts:19]: not table view specific

File: keith-vscode/src/table/actions.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+interface Action {
+    kind: string;
+}
+
+export interface AddRowAction extends Action {
```
Comment (Eddykasp) [inline keith-vscode/src/table/actions.ts:22]: what happens when rows of different lengths are added?

File: keith-vscode/src/table/actions.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            kind: KIND,
+            rowId: rowId
+        };
+    }
+
+    export function isThisAction(action: Action): action is SelectedRowAction {
+        return action.kind === SelectedRowAction.KIND;
+    }
+}
+
```
Comment (Eddykasp) [inline keith-vscode/src/table/actions.ts:116]: individual cells should also be selectable somehow

File: table-webview/src/actions.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+ * + Kiel University
+ *   + Department of Computer Science
+ *     + Real-Time and Embedded Systems Group
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
```
Comment (Eddykasp) [inline table-webview/src/actions.ts:16]: this file looks exactly like table/actions

File: table-webview/src/main.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        const table = document.getElementById(this.identifier + '_table');
+        if (table) {
+            const newTable = createTable(this.identifier, this.headers);
+            patch(table, newTable);
+        }
+    }
+
+}
+
+new Starter();
```
Comment (Eddykasp) [inline table-webview/src/main.ts:128]: why export Starter and call it here anyway?

**Review by Eddykasp — CHANGES_REQUESTED** [review]
please add docstrings!

File: .gitignore [hunk 8]
```diff
@@ -12,3 +12,5 @@ yarn-error.log
 **/package-backup.json
 **/gen-webpack.config.js
 .DS_Store
+
+table-webview/out/
```
Comment (NiklasRentzCAU) [inline .gitignore:16]: just `out` would match better here imo. Why does this build to another folder anyways? I am comparing this to the other packages that can be used as dependencies in [klighd-vscode](https://github.com/kieler/klighd-vscode/tree/main/packages), where that is not necessary either. The executed scripts also seem to be a little different here.

File: .vscode/launch.json [hunk 9]
```diff
@@ -8,11 +8,13 @@
...
+            ],
```
Comment (NiklasRentzCAU) [inline .vscode/launch.json:14]: Does not belong to the default extension launch config, maybe as a separate one for launching both extensions together in a combined workspace

File: keith-vscode/package.json [hunk 10]
```diff
@@ -561,8 +561,9 @@
     "lint": "eslint .",
     "build": "webpack --mode production",
     "watch": "webpack --watch",
-    "package": "vsce package --yarn -o keith-vscode.vsix",
-    "vscode:prepublish": "yarn run build"
+    "package": "vsce package --yarn -o keith-vscode.vsix && yarn run copy-webview",
```
Comment (NiklasRentzCAU) [inline keith-vscode/package.json:564]: If you define a valid `package` script in the webview package.json, calling that script on the top level (as should be done for building the system) will automatically include building the webview code.

File: keith-vscode/src/constants.ts (hunk 3 above)
Comment (NiklasRentzCAU) [reply keith-vscode/src/constants.ts:19]: i guess this is already part of the implementation for the model checker table view that was not excluded fully from this PR. Does not hurt anyone to already include it here.

File: keith-vscode/src/table/actions.ts [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+ *     + Real-Time and Embedded Systems Group
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+interface Action {
```
Comment (NiklasRentzCAU) [inline keith-vscode/src/table/actions.ts:18]: What kind of an action is this? Only used for the table view?
As this should be public API for reusing table views for different use cases, this should be fully documented on its behavior and usage. Also applies for most other public API here.

File: keith-vscode/src/table/table-webview.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
+/* eslint-disable prettier/prettier */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable no-promise-executor-return */
+/* eslint-disable no-console */
+/* eslint-disable no-return-assign */
+/* eslint-disable prefer-template */
+/* eslint-disable prettier/prettier */
```
Comment (NiklasRentzCAU) [inline keith-vscode/src/table/table-webview.ts:7]: We would like to use the configured eslint/prettier configuration, disabling should only be done for good (documented) reasons or the configuration should be changed if we really do not want to enforce some of the configured behavior.
Do you have the prettier extension installed in your VS Code? That allows to show the errors while writing the code, not only when trying to build it.

File: keith-vscode/src/table/table-webview.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        // TODO: headers should be able by sending the webview a msg
+        webview.html = `
+            <!DOCTYPE html>
+            <html lang="en">
+                <head>
+                    <meta charset="UTF-8">
+                    <meta name="viewport" content="width=device-width, height=device-height">
+                    <title>${title}</title>
+                    <link
+                        rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css"
```
Comment (NiklasRentzCAU) [inline keith-vscode/src/table/table-webview.ts:82]: We should not require an internet connection and link to an external web page for this to load properly - fontawesome has some nicer ways to be embedded into Typescript code.
Furthermore, I think we cannot force such external web page accesses due to DSGVO reasons.

File: keith-vscode/src/table/table-webview.ts [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * Handles messages from the webview.
+     * @param message The message received from the webview.
+     */
+    protected async receiveFromWebview(message: any) {
+        console.log("Received from template webview");
+        if (message.readyMessage) {
+            this.resolveWebviewReady();
+            this.sendTableIdentifier();
+        }else if (message.action) {
```
Comment (NiklasRentzCAU) [inline keith-vscode/src/table/table-webview.ts:144]: missing whitespace

File: table-webview/package.json [hunk 15]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "table-webview",
```
Comment (NiklasRentzCAU) [inline table-webview/package.json:2]: @kieler/table-webview

File: table-webview/package.json [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+  "version": "0.0.1",
+  "description": "Table as Webview",
+  "author": "Kiel University <rt-kieler-devel@informatik.uni-kiel.de>",
+  "license": "EPL-2.0",
+  "keywords": [
+    "vscode"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/kieler",
```
Comment (NiklasRentzCAU) [inline table-webview/package.json:12]: ...kieler/vscode?

File: table-webview/package.json [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    "file-loader": "^6.2.0",
+    "rimraf": "latest",
+    "source-map-loader": "^3.0.0",
+    "style-loader": "^1.2.1",
+    "ts-loader": "^9.2.6",
+    "typescript": "3.8.3",
+    "webpack": "^5.64.2",
+    "webpack-cli": "^4.9.1",
+    "snabbdom": "^3.5.1"
+  },
```
Comment (NiklasRentzCAU) [inline table-webview/package.json:34]: Please make sure that all dependencies that are required to run the code are listed in `dependencies`, as that lists everything needed to run the code. `dev-dependencies` is only for additional dependencies used to build the code (such as the *-loader packages, rimraf, etc.

File: table-webview/src/html.tsx [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+ *     + Real-Time and Embedded Systems Group
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+import { attributesModule, classModule, eventListenersModule, init, propsModule, styleModule, VNode, h, jsx } from 'snabbdom';
```
Comment (NiklasRentzCAU) [inline table-webview/src/html.tsx:18]: please sort the imports alphabetically

File: table-webview/src/main.ts (hunk 7 above)
Comment (NiklasRentzCAU) [reply table-webview/src/main.ts:128]: From what I understood this should be a generic webview that can be reused for any view. Then this generic code should not be started here, but extended/used and started by the views using this table view.

File: table-webview/src/mouseListener.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        if (lastSelected) {
+            lastSelected.classList.remove("focused");
+        }
+        lastSelected = owner;
+        const action = { kind: SelectedRowAction.KIND, rowId: owner.id } as SelectedRowAction;
+        owner.classList.add("focused");
+        return action;
+    }
+    return undefined;
+}
```
Comment (NiklasRentzCAU) [inline table-webview/src/mouseListener.ts:35]: Why is this no class? How should this be reusable for other webviews if there is only this one mouse listener?

File: table-webview/webpack.config.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                test: /\.(ttf)$/,
+                loader: 'file-loader',
+                options: {
+                    name: '[name].[ext]',
+                    outputPath: '',
+                    publicPath: '..',
+                    postTransformPublicPath: (p) => `__webpack_public_path__ + ${p}`,
+                }
+            },
```
Comment (NiklasRentzCAU) [inline table-webview/webpack.config.js:45]: This looks like not needed copypasta from another repository. Or do you intend to include own font files via .ttf?

File: table-webview/webpack.config.js [hunk 21]
```diff
@@ -0,0 +1,50 @@
+// @ts-check
```
Comment (NiklasRentzCAU) [inline table-webview/webpack.config.js:1]: Why do we need a webpack configuration if this should be generic code? I would have thought that any webview implementation reusing this table view code would pack it.

File: table-webview/package.json [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    "src",
+    "css"
+  ],
+  "dependencies": {
+    "reflect-metadata": "^0.1.13"
+  },
+  "devDependencies": {
+    "css-loader": "^6.5.1",
+    "file-loader": "^6.2.0",
+    "rimraf": "latest",
```
Comment (NiklasRentzCAU) [inline table-webview/package.json:26]: I think the community tries to always have a set minimum version instead of just a latest tag.

**Review by NiklasRentzCAU — CHANGES_REQUESTED** [review]
In general this code needs documentation on how to use it, just from the code I am not sure how I would implement a new view using this code. Please provide more documentation so we can better review this and understand how to use it.

Some detailed comments are in the code.

File: keith-vscode/src/table/actions.ts (hunk 4 above)
Comment (Drakae) [reply keith-vscode/src/table/actions.ts:22]: is no problem

File: table-webview/src/actions.ts (hunk 6 above)
Comment (Drakae) [reply table-webview/src/actions.ts:16]: i removed table/actions

File: table-webview/package.json [hunk 23]
```diff
@@ -9,33 +9,31 @@
...
-    "reflect-metadata": "^0.1.13"
+    "reflect-metadata": "^0.1.13",
+    "typescript": "3.8.3",
+    "snabbdom": "^3.5.1"
   },
   "devDependencies": {
     "css-loader": "^6.5.1",
     "file-loader": "^6.2.0",
-    "rimraf": "latest",
+    "rimraf": "^3.0.2",
```
Comment (Eddykasp) [inline table-webview/package.json:27]: where is this used? what is it for?

File: table-webview/src/table-webview.ts [hunk 24]
```diff
@@ -32,7 +25,7 @@ export class TableWebview {
     protected title: string;
```
Comment (Eddykasp) [inline table-webview/src/table-webview.ts:25]: docstrings

File: table-webview/src/table-webview.ts [hunk 25]
```diff
@@ -137,19 +170,23 @@ export class TableWebview {
      * @param message The message received from the webview.
      */
     protected async receiveFromWebview(message: any) {
-        console.log("Received from template webview");
+        console.log('Received from template webview')
```
Comment (Eddykasp) [inline table-webview/src/table-webview.ts:173]: necessary to log to console?

**Pull request closed without merging**